internal/admin/handle_health_authority.go

```diff
@@ -0,0 +1,240 @@
+		}
+
+		if action := c.Param("action"); action == "create" {
+			var form keyhealthAuthorityFormData
+			err := c.Bind(&form)
+			if err != nil {
+				ErrorPage(c, err.Error())
+				return
+			}
+
+			var hak model.HealthAuthorityKey
+			err = form.PopulateHealthAuthorityKey(&hak)
+			if err != nil {
+				ErrorPage(c, fmt.Sprintf("Error parsing new health authority key: %v", err))
+				return
+			}
+			err = haDB.AddHealthAuthorityKey(ctx, healthAuthority, &hak)
+			if err != nil {
+				ErrorPage(c, fmt.Sprintf("Error saving health authority key: %v", err))
+				return
+			}
+
+		} else if action == "revoke" || action == "reinstate" || action == "activate" {
+			version := c.Param("version")
+
+			// find the key.
+			var hak *model.HealthAuthorityKey
+			for _, key := range healthAuthority.Keys {
+				if key.Version == version {
+					hak = key
+					break
+				}
+			}
+			if hak == nil {
+				ErrorPage(c, "Invalid key specified")
+				return
+			}
+
+			if action == "activate" {
+				if hak.IsFuture() {
+					hak.From = time.Now()
+				}
+			} else if action == "revoke" {
+				hak.Revoke()
+			} else {
+				hak.Thru = time.Time{}
+			}
+
+			log.Printf("HAK %+v", hak)
+
+			err = haDB.UpdateHealthAuthorityKey(ctx, hak)
+			if err != nil {
+				ErrorPage(c, fmt.Sprintf("Error saving health authority key: %v", err))
+				return
+			}
+
+		} else {
+			ErrorPage(c, "invalid action")
+			return
+		}
+		c.Redirect(http.StatusFound, fmt.Sprintf("/healthauthority/%v", healthAuthority.ID))
+	}
+}
+
+type healthAuthorityFormData struct {
+	Issuer         string `form:"issuer"`
+	Audience       string `form:"audience"`
+	Name           string `form:"name"`
+	EnableStatsAPI bool   `form:"enable-stats-api"`
+	JwksURI        string `form:"jwks-uri"`
+}
+
+func (f *healthAuthorityFormData) PopulateHealthAuthority(ha *model.HealthAuthority) {
+	ha.Issuer = f.Issuer
+	ha.Audience = f.Audience
+	ha.Name = f.Name
+	ha.EnableStatsAPI = f.EnableStatsAPI
+	ha.SetJWKS(f.JwksURI)
+}
+
+type keyhealthAuthorityFormData struct {
+	Version  string `form:"version"`
+	PEMBlock string `form:"public-key-pem"`
+	FromDate string `form:"from-date"`
+	FromTime string `form:"from-time"`
+	ThruDate string `form:"Thru-date"`
+	ThruTime string `form:"thru-time"`
+}
+
+func (f *keyhealthAuthorityFormData) FromTimestamp() (time.Time, error) {
+	return CombineDateAndTime(f.FromDate, f.FromTime)
+}
+
+func (f *keyhealthAuthorityFormData) ThruTimestamp() (time.Time, error) {
+	return CombineDateAndTime(f.ThruDate, f.ThruTime)
+}
+
+func (f *keyhealthAuthorityFormData) PopulateHealthAuthorityKey(hak *model.HealthAuthorityKey) error {
+	fTime, err := f.FromTimestamp()
+	if err != nil {
+		return err
+	}
+	tTime, err := f.ThruTimestamp()
+	if err != nil {
+		return err
+	}
+	hak.Version = strings.TrimSpace(f.Version)
+	hak.From = fTime
+	hak.Thru = tTime
+	hak.PublicKeyPEM = strings.TrimSpace(f.PEMBlock)
+
+	_, err = hak.PublicKey()
+	if err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
Please simplify this check to
```go
return err
```

Reply:
This PR is just moving existing code around. To reduce the size of a diff and to reduce the chance of introducing a bug, I'm only moving code, no refactors or other changes in this PR.